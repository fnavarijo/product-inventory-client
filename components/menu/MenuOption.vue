<template>
  <div class="flex flex-col min-h-6 justify-start">
    <span
      class="p-1 mb-2 option rounded-md capitalize select-none cursor-pointer hover:bg-light-gray-700 hover:bg-opacity-50 transition duration-100 ease-linear"
      @click="displaySubItems"
    >
      {{ option }}
    </span>
    <ul
      class="h-auto relative overflow-hidden transition-all duration-100 ease-in-slide"
      :class="canDisplaySubItems ? 'flex-over-1' : 'flex-over-0'"
    >
      <transition name="slide-down">
        <div
          v-show="canDisplaySubItems"
          class="relative top-0"
        >
          <li
            v-for="(subItem, index) in subItems"
            :key="index"
            class="subitems pl-2 py-1 capitalize border-l-2 border-dark-gray-700"
          >
            <span class="py-1 px-2 rounded-md hover:bg-light-gray-700 hover:bg-opacity-50 transition duration-100 ease-linear">
              {{ subItem }}
            </span>
          </li>
        </div>
      </transition>
    </ul>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue';

export default Vue.extend({
  name: 'MenuOption',
  props: {
    option: {
      type: String,
      required: true,
    },
    subItems: {
      type: Array as PropType<Array<string>>,
      default: () => [],
    },
  },
  data () {
    return {
      canDisplaySubItems: false,
    };
  },
  methods: {
    displaySubItems (): void {
      this.canDisplaySubItems = !this.canDisplaySubItems;
    },
  },
});
</script>

<style scoped>
.flex-over-1 {
  flex: 1;
}

.flex-over-0 {
  flex: 0;
}

.slide-down-enter-active,
.slide-down-leave-active {
  @apply transition-top duration-200 ease-in-slide;
}

.slide-down-enter,
.slide-down-leave-to {
  @apply absolute -top-20;
}
</style>
