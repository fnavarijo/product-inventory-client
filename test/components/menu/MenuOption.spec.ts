import { mount } from '@vue/test-utils';

import MenuOption from '@/components/menu/MenuOption.vue';

describe('MenuOption component', () => {
  it('should display option without sublist', () => {
    const wrapper = mount(MenuOption, {
      propsData: {
        option: 'Base',
        subItems: ['Sub', 'Item'],
      },
    });

    expect(wrapper.find('.subitems').isVisible()).toBeFalsy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display sublist on click', async () => {
    const wrapper = mount(MenuOption, {
      propsData: {
        option: 'Base',
        subItems: ['Sub', 'Item'],
      },
    });

    await wrapper.find('.option').trigger('click');

    expect(wrapper.find('.subitems').isVisible()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
