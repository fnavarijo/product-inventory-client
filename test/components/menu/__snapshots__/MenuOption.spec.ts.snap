// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuOption component should display option without sublist 1`] = `
<div class="flex flex-col min-h-6 justify-start"><span class="p-1 mb-2 option rounded-md capitalize select-none cursor-pointer hover:bg-light-gray-600 hover:bg-opacity-50 transition duration-100 ease-linear">
    Base
  </span>
  <ul class="h-auto relative overflow-hidden transition-all duration-100 ease-in-slide flex-over-0">
    <transition-stub name="slide-down">
      <div class="relative top-0" style="display: none;">
        <li class="subitems pl-2 py-1 capitalize border-l border-dark-gray-700"><span class="py-1 px-2 rounded-md hover:bg-light-gray-600 hover:bg-opacity-50 transition duration-100 ease-linear">
            Sub
          </span></li>
        <li class="subitems pl-2 py-1 capitalize border-l border-dark-gray-700"><span class="py-1 px-2 rounded-md hover:bg-light-gray-600 hover:bg-opacity-50 transition duration-100 ease-linear">
            Item
          </span></li>
      </div>
    </transition-stub>
  </ul>
</div>
`;

exports[`MenuOption component should display sublist on click 1`] = `
<div class="flex flex-col min-h-6 justify-start"><span class="p-1 mb-2 option rounded-md capitalize select-none cursor-pointer hover:bg-light-gray-600 hover:bg-opacity-50 transition duration-100 ease-linear">
    Base
  </span>
  <ul class="h-auto relative overflow-hidden transition-all duration-100 ease-in-slide flex-over-1">
    <transition-stub name="slide-down">
      <div class="relative top-0" style="">
        <li class="subitems pl-2 py-1 capitalize border-l border-dark-gray-700"><span class="py-1 px-2 rounded-md hover:bg-light-gray-600 hover:bg-opacity-50 transition duration-100 ease-linear">
            Sub
          </span></li>
        <li class="subitems pl-2 py-1 capitalize border-l border-dark-gray-700"><span class="py-1 px-2 rounded-md hover:bg-light-gray-600 hover:bg-opacity-50 transition duration-100 ease-linear">
            Item
          </span></li>
      </div>
    </transition-stub>
  </ul>
</div>
`;
